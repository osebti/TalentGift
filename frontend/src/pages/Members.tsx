//#region import
// utilities
import { twMerge } from "tailwind-merge";
import { useState, useEffect, Dispatch, SetStateAction } from "react";
import { useGlobalState } from "@/state/global";
import config from "../../../config";

// components
import PageTitle from "@/components/PageTitle";
import Button from "@/components/Button";
import Modal from "@/components/Modal";
import { useNavigate } from "react-router-dom";
import { HiOutlineDocumentReport } from "react-icons/hi";
//#endregion

//#region types & interfaces
interface Member {
  name: string;
  id: string;
  membership: string;
  position: string;
}

interface GetMembersProps {
  oid: string;
}

interface MemberTableHeaderProps {
  columns: Member;
  style?: string;
}

interface MemberRowProps {
  member: Member;
  style?: string;
  btnStyle?: string;
  deleteRow: (id: string) => void;
  loggedInUserID: string;
}

interface MemberTableProps {
  members: Member[];
  setMembers: Dispatch<SetStateAction<Member[]>>;
  loggedInUserID: string;
}
//#endregion

//#region data handling

/**
 * This function sends a request to the server
 * to delete a member from an organization
 * @param id The ID of the member to delete
 */
function DeleteMember(id: string) {
  fetch(`http://${config.hostname}/members/${id}`, {
    method: "DELETE",
  });
}
//#endregion

//#region fragment component

/**
 * This renders the header for a members table
 * @param props Member Table Header Props
 * @return A header for the members table
 */
function MemberTableHeader(props: MemberTableHeaderProps) {
  const { columns, style } = props;
  return (
    <li
      id="table-header"
      className={twMerge(
        "flex flex-row w-full rounded-xl border-0 text-black items-center",
        style
      )}
    >
      <div className="w-1/5 px-4 md:px-8">{columns.name}</div>
      <div className="w-1/5 px-4 md:px-8">{columns.id}</div>
      <div className="w-1/5 px-4 md:px-8">{columns.membership}</div>
      <div className="w-1/5 px-4 md:px-8">{columns.position}</div>
      <div className="w-1/5 px-4 md:px-8">Options</div>
    </li>
  );
}

/**
 * This renders a row of the members table
 * @param props Member Row Props
 * @return A row of the members table containing a member's data
 */
function MemberRow(props: MemberRowProps) {
  const { member, style, btnStyle, deleteRow, loggedInUserID } = props;
  const navigate = useNavigate();
  return (
    <>
      <li
        id={member.id}
        className={twMerge(
          "relative w-full h-full flex flex-row rounded-xl border-0 text-black items-center",
          style
        )}
      >
        <div className="w-1/5 px-4 md:px-8">{member.name}</div>
        <div className="w-1/5 px-4 md:px-8">{member.id}</div>
        <div className="w-1/5 px-4 md:px-8">{member.membership}</div>
        <div className="w-1/5 px-4 md:px-8">{member.position}</div>
        {/* Do not allow users to delete themselves */}
        {member.id !== loggedInUserID ? (
          <div className="w-1/5 px-4 md:px-8 flex justify-around items-center gap-1">
            <button
              id={`${member.id}-view`}
              data-testid={`${member.id}-view`}
              className={twMerge("btn btn-square btn-outline", btnStyle)}
              onClick={() => navigate(`/dashboard/report/${member.id}`)}
            >
              <HiOutlineDocumentReport size={25} />
            </button>
            <button
              id={`${member.id}-delete`}
              data-testid={`${member.id}-delete`}
              className={twMerge("btn btn-square btn-outline", btnStyle)}
              onClick={() => {
                (
                  document.getElementById(
                    "delete-member-modal"
                  ) as HTMLDialogElement
                ).showModal();
                deleteRow(member.id);
              }}
            >
              <svg
                xmlns="http://www.w3.org/2000/svg"
                className="h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M6 18L18 6M6 6l12 12"
                />
              </svg>
            </button>
          </div>
        ) : null}
      </li>
    </>
  );
}

/**
 * This renders a table of the organization's members
 * @param props Member Table Props
 * @returns A members table
 */
function MemberTable(props: MemberTableProps) {
  const { members, setMembers, loggedInUserID } = props;
  const [selectedMemberID, setSelectedMemberID] = useState("0");
  return (
    <>
      <ul className="w-full flex flex-col justify-center items-center gap-4">
        {/** Row containing column names */}
        <MemberTableHeader
          columns={{
            name: "Name",
            id: "ID",
            membership: "Membership",
            position: "Position",
          }}
          style="min-h-[2rem] bg-neutralTransparent font-semibold"
        />
        {/** Rows containing member data */}
        {members.map((elem, index) => (
          <MemberRow
            key={index}
            member={elem}
            style={twMerge(
              "min-h-[4em]",
              index % 2 === 0 ? "bg-primaryTransparent" : "bg-accentTransparent"
            )}
            deleteRow={() => setSelectedMemberID(members[index].id)}
            btnStyle={index % 2 === 0 ? "btn-accent" : "btn-primary"}
            loggedInUserID={loggedInUserID}
          />
        ))}
      </ul>
      <Modal
        id="delete-member"
        header="Delete Member"
        message={`Are you sure you want to delete member with UID ${selectedMemberID}?`}
        onConfirm={() => {
          DeleteMember(selectedMemberID);
          const updatedMembers = members.filter(
            (member) => member.id != selectedMemberID
          );
          setMembers(updatedMembers);
        }}
      />
    </>
  );
}

//#endregion

export default function Members() {
  const initialArr: Member[] = []; // Create an empty array of type Member[]
  const [members, setMembers] = useState(initialArr);
  // Problem: mid and oid become undefined if the user refreshes the page

  const [user] = useGlobalState("user");
  const [org] = useGlobalState("org");

  const oid = org.oid;
  const uid = user.uid;

  useEffect(() => {
    async function GetMembers(props: GetMembersProps) {
      const { oid } = props;
      const membersResponse = await fetch(
        `http://${config.hostname}/organization/members/${oid}`
      );
      const membersData = await membersResponse.json();
      const membersArr: Member[] = [];
      membersData.forEach(
        (memberData: {
          firstname: string;
          lastname: string;
          uid: number;
          membership: string;
          role: string;
        }) => {
          membersArr.push({
            name: `${memberData.firstname} ${memberData.lastname}`,
            id: String(memberData.uid),
            membership: memberData.membership,
            position: memberData.role,
          });
        }
      );
      setMembers(membersArr);
    }

    GetMembers({ oid }).catch(console.error);
  }, [members.length, uid, oid]);

  return (
    <div className="w-full flex flex-col justify-center gap-12 px-8 md:px-12 pb-8">
      {/** Title of page */}
      <PageTitle id="members" title="Members" />
      {/** Add members button (directs to AddMembers page) */}
      <Button
        id="add-members"
        label="Add Members"
        href="/dashboard/add-members"
        navigate
        style="btn-secondary rounded-2xl"
      />
      {/** Table listing members */}
      <MemberTable
        members={members}
        setMembers={setMembers}
        loggedInUserID={uid}
      />
    </div>
  );
}
