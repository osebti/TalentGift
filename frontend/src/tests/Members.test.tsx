//#region imports
import { cleanup, fireEvent, render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import fetchMock from "jest-fetch-mock";

import Members from "@/pages/Members";
//#endregion

//#region test setup
fetchMock.enableMocks();
console.error = jest.fn();
console.warn = jest.fn();
console.log = jest.fn();

beforeEach(() => {
  fetchMock.resetMocks();
});

afterEach(cleanup);
//#endregion

//#region functions and objects used for testing
const mockMembers = [
  {
    firstname: "John",
    lastname: "Smith",
    uid: 1,
    membership: "Manager",
    role: "CEO",
  },
  {
    firstname: "Erika",
    lastname: "Lee",
    uid: 2,
    membership: "Member",
    role: "Accountant",
  },
];

function renderMembersPage() {
  // mock fetch request for members
  fetchMock.mockResponseOnce(JSON.stringify(mockMembers), {
    status: 200,
    headers: { "Content-Type": "application/json" },
  });

  // render page
  const membersPage = render(
    <BrowserRouter>
      <Members />
    </BrowserRouter>
  );

  // check to ensure fetch was called
  expect(fetchMock).toHaveBeenCalledTimes(1);

  return membersPage;
}
//#endregion

//#region tests
describe("Rendering the Members Page", () => {
  it("renders correctly", async () => {
    const page = renderMembersPage();
    await waitFor(() => {
      expect(page).toMatchSnapshot();
    });
  });

  it("renders the add members button", async () => {
    const { queryByTestId } = renderMembersPage();
    await waitFor(() => {
      expect(queryByTestId("add-members-button")).not.toBeNull();
    });
  });
});

describe("Displaying the Organization Members", () => {
  it("displays all of the organization members", async () => {
    const { queryByText } = renderMembersPage();
    await waitFor(() => {
      expect(queryByText("John Smith")).not.toBeNull();
      expect(queryByText("Erika Lee")).not.toBeNull();
    });
  });
});

describe("Deleting the Organization Members", () => {
  it("member is deleted from members table if the user confirms", async () => {
    const { queryByTestId, queryByText } = renderMembersPage();

    await waitFor(() => {
      // click the delete button for member Erika
      const deleteButton = queryByTestId("2-delete");
      if (deleteButton != null) {
        fireEvent.click(deleteButton);
      }
      // click the confirm button
      const confirmButton = queryByTestId("Confirm-button");
      if (confirmButton != null) {
        fireEvent.submit(confirmButton);
      }
      // check member deletion
      expect(queryByText("Erika Lee")).toBeNull();
    });
  });

  it("member is not deleted from members table if the user cancels", async () => {
    const { queryByTestId, queryByText } = renderMembersPage();

    // click the delete button for member Erika
    await waitFor(() => {
      // click the delete button for member Erika
      const deleteButton = queryByTestId("2-delete");
      if (deleteButton != null) {
        fireEvent.submit(deleteButton);
      }
      // click the cancel button
      const cancelButton = queryByTestId("Cancel-button");
      if (cancelButton != null) {
        fireEvent.submit(cancelButton);
      }
      // check for member
      expect(queryByText("Erika Lee")).not.toBeNull();
    });
  });

});

describe("Viewing Member Reports", () => {

  it("the view report button directs to the employee report", async () => {
    const { queryByTestId } = renderMembersPage();

    await waitFor(() => {
      // click the view button for member Erika
      const viewButton = queryByTestId("2-view");
      if (viewButton != null) {
        fireEvent.click(viewButton);
      }
      // check for redirect to her report
      expect(global.window.location.href).toContain("/dashboard/report/2");
    });

  });

});
//#endregion
