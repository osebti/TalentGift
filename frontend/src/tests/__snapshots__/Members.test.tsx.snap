// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the Members Page renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full flex flex-col justify-center gap-12 px-8 md:px-12 pb-8"
      >
        <div
          class="w-full font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
          data-testid="members-page-title"
          id="members-page-title"
        >
          Members
        </div>
        <a
          class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary rounded-2xl"
          data-testid="add-members-button"
          href="/dashboard/add-members"
          id="add-members-button"
        >
          Add Members
        </a>
        <ul
          class="w-full flex flex-col justify-center items-center gap-4"
        >
          <li
            class="flex flex-row w-full rounded-xl border-0 text-black items-center min-h-[2rem] bg-neutralTransparent font-semibold"
            id="table-header"
          >
            <div
              class="w-1/5 px-4 md:px-8"
            >
              Name
            </div>
            <div
              class="w-1/5 px-4 md:px-8"
            >
              ID
            </div>
            <div
              class="w-1/5 px-4 md:px-8"
            >
              Membership
            </div>
            <div
              class="w-1/5 px-4 md:px-8"
            >
              Position
            </div>
            <div
              class="w-1/5 px-4 md:px-8"
            >
              Options
            </div>
          </li>
        </ul>
        <dialog
          class="modal"
          data-testid="delete-member-modal"
          id="delete-member-modal"
        >
          <div
            class="modal-box border-2 border-primary"
          >
            <h3
              class="font-bold text-lg"
              id="modal-header"
            >
              Delete Member
            </h3>
            <p
              class="py-4"
              id="modal-message"
            >
              Are you sure you want to delete member with UID 0?
            </p>
            <div
              class="modal-action"
            >
              <form
                method="dialog"
              >
                <div
                  class="flex gap-3"
                  id="modal-button-container"
                >
                  <button
                    class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-accent rounded-xl"
                    data-testid="Cancel-button"
                    id="Cancel-button"
                  >
                    Cancel
                  </button>
                  <button
                    class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-secondary rounded-xl"
                    data-testid="Confirm-button"
                    id="Confirm-button"
                  >
                    Confirm
                  </button>
                </div>
              </form>
            </div>
          </div>
        </dialog>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full flex flex-col justify-center gap-12 px-8 md:px-12 pb-8"
    >
      <div
        class="w-full font-semibold text-2xl md:text-3xl lg:text-4xl text-black"
        data-testid="members-page-title"
        id="members-page-title"
      >
        Members
      </div>
      <a
        class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn-secondary rounded-2xl"
        data-testid="add-members-button"
        href="/dashboard/add-members"
        id="add-members-button"
      >
        Add Members
      </a>
      <ul
        class="w-full flex flex-col justify-center items-center gap-4"
      >
        <li
          class="flex flex-row w-full rounded-xl border-0 text-black items-center min-h-[2rem] bg-neutralTransparent font-semibold"
          id="table-header"
        >
          <div
            class="w-1/5 px-4 md:px-8"
          >
            Name
          </div>
          <div
            class="w-1/5 px-4 md:px-8"
          >
            ID
          </div>
          <div
            class="w-1/5 px-4 md:px-8"
          >
            Membership
          </div>
          <div
            class="w-1/5 px-4 md:px-8"
          >
            Position
          </div>
          <div
            class="w-1/5 px-4 md:px-8"
          >
            Options
          </div>
        </li>
      </ul>
      <dialog
        class="modal"
        data-testid="delete-member-modal"
        id="delete-member-modal"
      >
        <div
          class="modal-box border-2 border-primary"
        >
          <h3
            class="font-bold text-lg"
            id="modal-header"
          >
            Delete Member
          </h3>
          <p
            class="py-4"
            id="modal-message"
          >
            Are you sure you want to delete member with UID 0?
          </p>
          <div
            class="modal-action"
          >
            <form
              method="dialog"
            >
              <div
                class="flex gap-3"
                id="modal-button-container"
              >
                <button
                  class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-accent rounded-xl"
                  data-testid="Cancel-button"
                  id="Cancel-button"
                >
                  Cancel
                </button>
                <button
                  class="btn btn-md capitalize font-normal text-white text-sm md:text-base btn btn-outline btn-secondary rounded-xl"
                  data-testid="Confirm-button"
                  id="Confirm-button"
                >
                  Confirm
                </button>
              </div>
            </form>
          </div>
        </div>
      </dialog>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
